src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultHtmlGenerator.cs

```diff
@@ -422,16 +422,21 @@ public virtual TagBuilder GenerateAntiForgery([NotNull] ViewContext viewContext)
             var type = allowMultiple ? typeof(string[]) : typeof(string);
             var defaultValue = GetModelStateValue(viewContext, fullName, type);
 
-            // If we haven't already used ViewData to get the entire list of items then we need to
-            // use the ViewData-supplied value before using the parameter-supplied value.
-            if (defaultValue == null && !string.IsNullOrEmpty(expression))
+            // If ModelState did not contain a current value, fall back to ViewData- or ModelExplorer-supplied value.
+            if (defaultValue == null)
             {
-                if (!usedViewData)
+                if (modelExplorer == null)
                 {
-                    defaultValue = viewContext.ViewData.Eval(expression);
+                    // Html.DropDownList() and Html.ListBox() helper case.
+                    // Cannot use ViewData if it contains the select list.
+                    if (!usedViewData)
+                    {
+                        defaultValue = viewContext.ViewData.Eval(expression);
+                    }
                 }
-                else if (modelExplorer != null)
+                else
                 {
+                    // <select/>, Html.DropDownListFor() and Html.ListBoxFor() helper case. Do not use ViewData.
```

Comment:
Comments like this are really helpful for other that have to work on this code from time to time. :100: